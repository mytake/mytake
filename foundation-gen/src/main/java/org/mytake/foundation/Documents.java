/*
 * MyTake.org transcript GUI.
 * Copyright (C) 2017-2020 MyTake.org, Inc.
 * 
 * The MyTake.org transcript GUI is licensed under EPLv2
 * because SWT is incompatible with AGPLv3, the rest of
 * MyTake.org is licensed under AGPLv3.
 *
 * This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License 2.0
 * which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-2.0/
 */
package org.mytake.foundation;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.security.NoSuchAlgorithmException;
import java2ts.FT;
import java2ts.FT.DocumentFactContent;
import java2ts.FT.Fact;
import org.mytake.foundation.parsers.FoundationParser;

public class Documents {
	public static void national(FactWriter writer) throws NoSuchAlgorithmException, IOException {
		Documents documents = new Documents(writer);
		documents.add("United States Constitution", "1788-06-21");
		documents.add("Bill of Rights", "1791-12-15");
		documents.add("Amendment 11", "1795-02-07");
		documents.add("Amendment 12", "1804-06-15");
		documents.add("Amendment 13", "1865-12-06");
		documents.add("Amendment 14", "1868-07-09");
		documents.add("Amendment 15", "1870-02-03");
		documents.add("Amendment 16", "1913-02-03");
		documents.add("Amendment 17", "1913-04-08");
		documents.add("Amendment 18", "1919-01-16");
		documents.add("Amendment 19", "1920-08-18");
		documents.add("Amendment 20", "1933-01-23");
		documents.add("Amendment 21", "1933-12-05");
		documents.add("Amendment 22", "1951-02-27");
		documents.add("Amendment 23", "1961-03-29");
		documents.add("Amendment 24", "1964-01-23");
		documents.add("Amendment 25", "1967-02-10");
		documents.add("Amendment 26", "1971-07-01");
		documents.add("Amendment 27", "1992-05-05");
	}

	private final FactWriter writer;

	private Documents(FactWriter writer) {
		this.writer = writer;
	}

	private void add(String title, String date) throws NoSuchAlgorithmException, IOException {
		add(title, date, "ratified", FT.KIND_DOCUMENT);
	}

	private void add(String title, String date, String dateKind, String kind) throws NoSuchAlgorithmException, IOException {
		DocumentFactContent content = new DocumentFactContent();
		content.fact = new Fact();
		content.fact.title = title;
		content.fact.primaryDate = date;
		content.fact.primaryDateKind = dateKind;
		content.fact.kind = kind;

		// load the components
		String input = read(FactWriter.slugify(content.fact.title) + ".foundation.html");
		content.components = FoundationParser.toComponents(input);

		writer.writeAndReturnHash(content);
	}

	private String read(String name) throws IOException {
		return new String(Files.readAllBytes(Folders.SRC_DOCUMENT.resolve(name)), StandardCharsets.UTF_8);
	}
}
