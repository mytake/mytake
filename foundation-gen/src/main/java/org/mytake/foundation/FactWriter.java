/*
 * MyTake.org transcript GUI.
 * Copyright (C) 2017-2020 MyTake.org, Inc.
 * 
 * The MyTake.org transcript GUI is licensed under EPLv2
 * because SWT is incompatible with AGPLv3, the rest of
 * MyTake.org is licensed under AGPLv3.
 *
 * This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License 2.0
 * which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-2.0/
 */
package org.mytake.foundation;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.security.NoSuchAlgorithmException;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java2ts.FT;
import java2ts.FT.FactLink;

public class FactWriter {
	final List<FactLink> factLinks = new ArrayList<>();
	final Path dstDir;

	public FactWriter(Path dstDir) throws IOException {
		this.dstDir = dstDir;
		Files.createDirectories(dstDir);
	}

	protected static String slugify(String input) {
		return input.toLowerCase(Locale.ROOT)
				.replace(' ', '-') // replace spaces with hyphens
				.replaceAll("[-]+", "-") // replace multiple hypens with a single hyphen
				.replaceAll("[^\\w-]+", ""); // replace non-alphanumerics and non-hyphens
	}

	public String writeAndReturnHash(FT.FactContent content) throws IOException, NoSuchAlgorithmException {
		Hashed hashed = Hashed.asJson(content);
		Files.write(dstDir.resolve(hashed.hash + ".json"), hashed.content);

		FactLink link = new FactLink();
		link.fact = content.fact;
		link.hash = hashed.hash;
		factLinks.add(link);
		return link.hash;
	}

	public Hashed hashedFactLinks() throws NoSuchAlgorithmException {
		return Hashed.asJson(factLinks);
	}
}
