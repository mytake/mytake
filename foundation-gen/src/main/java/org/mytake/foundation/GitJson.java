/*
 * MyTake.org transcript GUI.
 * Copyright (C) 2020 MyTake.org, Inc.
 * 
 * The MyTake.org transcript GUI is licensed under EPLv2
 * because SWT is incompatible with AGPLv3, the rest of
 * MyTake.org is licensed under AGPLv3.
 *
 * This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License 2.0
 * which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-2.0/
 */
package org.mytake.foundation;

import com.diffplug.common.base.Preconditions;

public class GitJson {
	public static final char COMMENT_OPEN = '⌊';
	public static final char COMMENT_CLOSE = '⌋';

	/**
	 * Minified json is guaranteed to have no newlines, just a long single-line string.
	 * This makes it terrible to git diff with a standard line-based differ.
	 * To improve that behavior:
	 * - add newlines anywhere, they will be removed
	 * - add ⌊⌋ pairs (mathematic floor) anywhere, they are treated as a comment and removed
	 */
	public static String recondense(String in, StringBuilder buffer) {
		Preconditions.checkArgument(!in.isEmpty());
		Preconditions.checkArgument(in.charAt(0) != '\n');
		Preconditions.checkArgument(in.charAt(0) != COMMENT_OPEN);
		buffer.setLength(0);
		int start = 0;
		int next;
		while ((next = nextParsePaint(in, start)) != 0) {
			if (next > 0) {
				// was '\n'
				buffer.append(in, start, next);
				start = next + 1; // skip the \n
			} else {
				// was ⌊, ignore everything up to the next ⌋
				buffer.append(in, start, -next);
				int close = in.indexOf(COMMENT_CLOSE, -next + 1);
				if (close == -1) {
					return buffer.toString();
				} else {
					start = close + 1;
				}
			}
		}
		if (start < in.length()) {
			buffer.append(in, start, in.length());
		}
		return buffer.toString();
	}

	private static int nextParsePaint(String in, int startFrom) {
		for (int i = startFrom; i < in.length(); ++i) {
			char c = in.charAt(i);
			if (c == '\n') {
				return i;
			} else if (c == COMMENT_OPEN) {
				return -i;
			}
		}
		return 0; // signifies end of string
	}

	public static String recondense(String in) {
		return recondense(in, new StringBuilder(in.length()));
	}
}
