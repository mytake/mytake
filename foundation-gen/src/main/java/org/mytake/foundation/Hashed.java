/*
 * MyTake.org transcript GUI.
 * Copyright (C) 2018 MyTake.org, Inc.
 * 
 * The MyTake.org transcript GUI is licensed under EPLv2
 * because SWT is incompatible with AGPLv3, the rest of
 * MyTake.org is licensed under AGPLv3.
 *
 * This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License 2.0
 * which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-2.0/
 */
package org.mytake.foundation;

import java.io.ByteArrayOutputStream;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.Base64;

public class Hashed {
	public final byte[] content;
	public final String hash;

	private Hashed(byte[] content, String hash) {
		this.content = content;
		this.hash = hash;
	}

	public static Hashed asJson(Object content) throws NoSuchAlgorithmException {
		ByteArrayOutputStream output = new ByteArrayOutputStream();
		JsonMisc.toJson(content, output);
		byte[] contentBytes = output.toByteArray();
		MessageDigest digest = MessageDigest.getInstance("SHA-256");
		byte[] hash = digest.digest(contentBytes);

		String hashStr = Base64.getUrlEncoder().encodeToString(hash);
		return new Hashed(contentBytes, hashStr);
	}
}
