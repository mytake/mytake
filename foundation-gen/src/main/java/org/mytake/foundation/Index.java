/*
 * MyTake.org transcript GUI.
 * Copyright (C) 2017-2020 MyTake.org, Inc.
 * 
 * The MyTake.org transcript GUI is licensed under EPLv2
 * because SWT is incompatible with AGPLv3, the rest of
 * MyTake.org is licensed under AGPLv3.
 *
 * This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License 2.0
 * which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-2.0/
 */
package org.mytake.foundation;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java2ts.FT.IndexPointer;
import java2ts.Routes;
import org.mytake.lucene.Lucene;
import org.mytake.lucene.ZipMisc;

public class Index {
	public static void main(String[] args) throws Throwable {
		cleanDir(Folders.DST_FOUNDATION_DATA);
		Path luceneTemp = Files.createTempDirectory("mytake-lucene");

		FactWriter factWriter = new FactWriter(Folders.DST_FOUNDATION_DATA);
		Documents.national(factWriter);
		try (Lucene.Writer writer = new Lucene.Writer(luceneTemp)) {
			Videos.presidentialDebates(factContentJava -> {
				String hash = factWriter.writeAndReturnHash(factContentJava.toEncoded());
				writer.writeVideo(hash, factContentJava);
			});
		}

		// zip the index into a file, and delete the spread-out index
		ZipMisc.zip(luceneTemp, Folders.DST_FOUNDATION_RESOURCES.resolve(Lucene.INDEX_ARCHIVE));
		ZipMisc.deleteDir(luceneTemp);

		// write out the index
		Hashed hashed = factWriter.hashedFactLinks();
		Files.write(Folders.DST_FOUNDATION_DATA.resolve(hashed.hash + ".json"), hashed.content);

		// and the index-pointer
		IndexPointer pointer = new IndexPointer();
		pointer.hash = hashed.hash;
		// substring(1) to remove '/'
		Files.write(Folders.DST_FOUNDATION_DATA.getParent().resolve(Routes.FOUNDATION_INDEX_HASH.substring(1)), Hashed.asJson(pointer).content);
	}

	private static void cleanDir(Path folder) throws IOException {
		ZipMisc.deleteDir(folder);
		Files.createDirectories(folder);
	}
}
