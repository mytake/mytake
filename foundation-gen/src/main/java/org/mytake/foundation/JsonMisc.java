/*
 * MyTake.org transcript GUI. 
 * Copyright (C) 2017-2020 MyTake.org, Inc.
 * 
 * The MyTake.org transcript GUI is licensed under EPLv2
 * because SWT is incompatible with AGPLv3, the rest of
 * MyTake.org is licensed under AGPLv3.
 *
 * This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License 2.0
 * which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-2.0/
 */
package org.mytake.foundation;

import com.jsoniter.JsonIterator;
import com.jsoniter.output.EncodingMode;
import com.jsoniter.output.JsonStream;
import com.jsoniter.spi.Config;
import com.jsoniter.spi.DecodingMode;
import com.jsoniter.spi.JsoniterSpi;
import java.io.File;
import java.io.IOException;
import java.io.OutputStream;
import java.nio.file.Files;

public class JsonMisc {
	static {
		JsoniterSpi.registerTypeDecoder(Number.class, iter -> iter.readInt());
		CONFIG = new Config.Builder()
				.escapeUnicode(false)
				.decodingMode(DecodingMode.REFLECTION_MODE)
				.encodingMode(EncodingMode.REFLECTION_MODE)
				.build();
	}

	private static final Config CONFIG;

	public static void toJson(Object object, OutputStream output) {
		JsonStream.serialize(CONFIG, object, output);
	}

	public static String toJson(Object object) {
		return JsonStream.serialize(CONFIG, object);
	}

	public static <T> T fromJson(File file, Class<T> clazz) throws IOException {
		return fromJson(Files.readAllBytes(file.toPath()), clazz);
	}

	public static <T> T fromJson(byte[] content, Class<T> clazz) throws IOException {
		return JsonIterator.deserialize(CONFIG, content, clazz);
	}
}
