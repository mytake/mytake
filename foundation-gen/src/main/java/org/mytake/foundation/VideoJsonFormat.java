/*
 * MyTake.org transcript GUI. 
 * Copyright (C) 2020 MyTake.org, Inc.
 * 
 * The MyTake.org transcript GUI is licensed under EPLv2
 * because SWT is incompatible with AGPLv3, the rest of
 * MyTake.org is licensed under AGPLv3.
 *
 * This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License 2.0
 * which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-2.0/
 */
package org.mytake.foundation;

import com.diffplug.common.io.ByteStreams;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonParser;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.util.Comparator;
import java.util.List;
import java.util.regex.Pattern;
import java2ts.Foundation.Fact;
import java2ts.Foundation.Speaker;
import java2ts.Foundation.VideoFactMeta;

/** Format-friendly version of {@link VideoFactMeta#VideoFactMeta}. */
public class VideoJsonFormat {
	public Fact fact;
	public String youtubeId;
	public Number durationSeconds;
	public List<Speaker> speakers;

	public static void main(String[] args) throws IOException {
		byte[] content = ByteStreams.toByteArray(System.in);
		String formatted = format(new String(content, StandardCharsets.UTF_8));
		System.out.write(formatted.getBytes(StandardCharsets.UTF_8));
	}

	static String format(String input) throws IOException {
		VideoJsonFormat meta = JsonMisc.fromJson(input.getBytes(StandardCharsets.UTF_8), VideoJsonFormat.class);
		meta.speakers.sort(Comparator.comparing(speaker -> speaker.fullName));
		String result = JsonMisc.toJson(meta);

		Gson gson = new GsonBuilder().setPrettyPrinting().create();
		String formatted = gson.toJson(JsonParser.parseString(result));
		return SPEAKER.matcher(formatted).replaceAll("{\"fullName\": \"$1\", \"role\": \"$2\"}");
	}

	private static final Pattern SPEAKER = Pattern.compile("\\{\\s*\"fullName\": \"(.+)\",\\s+\"role\": \"(.+)\"\\s+\\}");
}
