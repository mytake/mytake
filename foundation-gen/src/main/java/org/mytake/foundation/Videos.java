/*
 * MyTake.org transcript GUI.
 * Copyright (C) 2018 MyTake.org, Inc.
 * 
 * The MyTake.org transcript GUI is licensed under EPLv2
 * because SWT is incompatible with AGPLv3, the rest of
 * MyTake.org is licensed under AGPLv3.
 *
 * This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License 2.0
 * which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-2.0/
 */
package org.mytake.foundation;

import com.diffplug.common.base.Throwing;
import compat.java2ts.VideoFactContentJava;
import java2ts.Foundation.VideoFactMeta;
import org.mytake.foundation.transcript.TranscriptFolder;
import org.mytake.foundation.transcript.TranscriptMatch;

public class Videos {
	public static void presidentialDebates(Throwing.Consumer<VideoFactContentJava> consumer) throws Throwable {
		TranscriptFolder folder = new TranscriptFolder(Folders.SRC_PRESIDENTIAL_DEBATES.toFile());
		for (String transcript : folder.transcriptsWithMeta()) {
			VideoFactContentJava factContent;
			try {
				TranscriptMatch match = folder.loadTranscript(transcript);
				factContent = match.toVideoFact();
			} catch (Exception e) {
				VideoFactMeta meta = folder.loadMetaNoValidation(transcript);
				VideoFactContentJava empty = new VideoFactContentJava();
				empty.fact = meta.fact;
				empty.youtubeId = meta.youtubeId;
				empty.durationSeconds = meta.durationSeconds.doubleValue();
				empty.speakers = meta.speakers;

				empty.plainText = "";
				empty.charOffsets = new int[0];
				empty.timestamps = new double[0];
				empty.speakerPerson = empty.charOffsets;
				empty.speakerWord = empty.charOffsets;
				factContent = empty;
			}
			consumer.accept(factContent);
		}
	}
}
