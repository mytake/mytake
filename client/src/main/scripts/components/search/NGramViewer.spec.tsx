/*
 * MyTake.org website and tooling.
 * Copyright (C) 2020 MyTake.org, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * You can contact us at team@mytake.org
 */
import React from "react";
import renderer from "react-test-renderer";
import { Foundation } from "../../common/foundation";
import NGramViewer, { _getSearchTerms } from "./NGramViewer";
import { SearchMode, _searchImpl, _SearchWithData } from "./search";
// this is created by LucenePlayground
import socialSecuritySearchResults from "./testData/socialSecuritySearchResults.json";

test("NGramViewer parses search terms", () => {
  const searchQueries = [
    "Wall, -Wall Street",
    "global warming, climate change",
  ];
  const expectedTerms = [["wall"], ["global warming", "climate change"]];

  searchQueries.forEach((q, idx) => {
    const terms = _getSearchTerms(q);
    expect(terms).toEqual(expectedTerms[idx]);
  });
});

test("NGramViewer social security", async () => {
  const result = _searchImpl(
    new _SearchWithData(
      "social security",
      socialSecuritySearchResults.facts,
      await Foundation.fetchAll(
        socialSecuritySearchResults.facts.map((f) => f.hash)
      ),
      SearchMode.BeforeAndAfter
    )
  );

  const tree = renderer
    .create(<NGramViewer searchResult={result} onBarClick={jest.fn()} />)
    .toJSON();

  expect(tree).toMatchSnapshot();
});
