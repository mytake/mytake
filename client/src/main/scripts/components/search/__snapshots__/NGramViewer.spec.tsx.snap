// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NGramViewer social security 1`] = `
<div
  className="ngram__outer-container"
>
  <div
    className="ngram__inner-container"
  >
    <svg
      height={268}
      width={700}
    />
    <div
      className="ngram__term-list"
    >
      <span
        className="ngram__term"
        style={
          Object {
            "color": "rgb(93, 177, 151)",
          }
        }
      >
        social security
      </span>
    </div>
  </div>
</div>
`;
